crates/qasm2/src/lex.rs

```diff
@@ -0,0 +1,772 @@
+    /// of the position, and the `done` state of the iterator.
+    fn advance_line(&mut self) -> PyResult<usize> {
+        if self.done {
+            Ok(0)
+        } else {
+            self.line += 1;
+            self.col = 0;
+            self.line_buffer.clear();
+            // We can assume that nobody's running this on ancient Mac software that uses only '\r'
+            // as its linebreak character.
+            match self.source.read_until(b'\n', &mut self.line_buffer) {
+                Ok(count) => {
+                    if count == 0 || self.line_buffer[count - 1] != b'\n' {
+                        self.done = true;
+                    }
+                    Ok(count)
+                }
+                Err(err) => {
+                    self.done = true;
+                    Err(QASM2ParseError::new_err(message_generic(
+                        Some(&Position::new(&self.filename, self.line, self.col)),
+                        &format!("lexer failed to read stream: {}", err),
+                    )))
+                }
+            }
+        }
+    }
+
+    /// Get the next character in the stream.  This updates the line and column information for the
+    /// current bute as well.
+    fn next_byte(&mut self) -> PyResult<Option<u8>> {
+        if self.col >= self.line_buffer.len() && self.advance_line()? == 0 {
+            return Ok(None);
+        }
+        let out = self.line_buffer[self.col];
+        self.col += 1;
+        match out {
+            b @ 0x80..=0xff => {
+                self.done = true;
+                Err(QASM2ParseError::new_err(message_generic(
+                    Some(&Position::new(&self.filename, self.line, self.col)),
+                    &format!("encountered a non-ASCII byte: {:02X?}", b),
+                )))
+            }
+            b => Ok(Some(b)),
+        }
+    }
+
+    /// Peek at the next byte in the stream without consuming it.  This still returns an error if
+    /// the next byte isn't in the valid range for OpenQASM 2, or if the file/stream has failed to
+    /// read into the buffer for some reason.
+    fn peek_byte(&mut self) -> PyResult<Option<u8>> {
+        if self.col >= self.line_buffer.len() && self.advance_line()? == 0 {
+            return Ok(None);
+        }
+        match self.line_buffer[self.col] {
+            b @ 0x80..=0xff => {
+                self.done = true;
+                Err(QASM2ParseError::new_err(message_generic(
+                    Some(&Position::new(&self.filename, self.line, self.col)),
+                    &format!("encountered a non-ASCII byte: {:02X?}", b),
+                )))
+            }
+            b => Ok(Some(b)),
+        }
+    }
+
+    /// Expect that the next byte is not a word continuation, providing a suitable error message if
+    /// it is.
+    fn expect_word_boundary(&mut self, after: &str, start_col: usize) -> PyResult<()> {
```

Comment:
They are, but that's deliberate - it's unified error handling, rather than having an `is_word_boundary` function followed by the same error handling everywhere it's called.  I can change it to that second form if you think it would be better.